### ISSUE TITLE
Area code not set

### ISSUE THREAD

**Issue by juliangardnereurocios** [Smile-SA/elasticsuite#334]
Hi,

after installing the elastic suite through composer, I get the following error when I try the upgrade:

```
php bin/magento setup:upgrade

Cache cleared successfully
File system cleanup:
var/generation/CloudFlare
var/generation/Composer
var/generation/J2t
var/generation/Magento
var/generation/Smile
var/generation/Symfony
The directory 'var/di/' doesn't exist - skipping cleanup
Updating modules:
Schema creation/updates:
Module 'Magento_Store':
Module 'CloudFlare_Plugin':
Module 'Magento_Directory':
Module 'Magento_Eav':
Module 'Magento_AdvancedPricingImportExport':
Module 'Magento_Backend':
Module 'Magento_Theme':
Module 'Magento_Authorization':
Module 'Magento_Backup':
Module 'Magento_Customer':
Module 'Magento_AdminNotification':
Module 'Magento_BundleImportExport':
Module 'Magento_CacheInvalidate':
Module 'Magento_Indexer':
Module 'Magento_Cms':
Module 'Magento_CatalogImportExport':
Module 'Magento_Catalog':
Module 'Magento_Rule':
Module 'Magento_Msrp':
Module 'Magento_Search':
Module 'Magento_CatalogUrlRewrite':
Module 'Magento_Widget':
Module 'Magento_Quote':
Module 'Magento_CheckoutAgreements':
Module 'Magento_SalesSequence':
Module 'Magento_CmsUrlRewrite':
Module 'Magento_Config':
Module 'Magento_ConfigurableImportExport':
Module 'Magento_Payment':
Module 'Magento_Contact':
Module 'Magento_Cookie':
Module 'Magento_Cron':
Module 'Magento_CurrencySymbol':
Module 'Magento_CatalogInventory':
Module 'Magento_CustomerImportExport':
Module 'Magento_Deploy':
Module 'Magento_Developer':
Module 'Magento_Dhl':
Module 'Magento_Sales':
Module 'Magento_Downloadable':
Module 'Magento_ImportExport':
Module 'Magento_Bundle':
Module 'Magento_Email':
Module 'Magento_User':
Module 'Magento_Fedex':
Module 'Magento_GiftMessage':
Module 'Magento_Checkout':
Module 'Magento_GoogleAnalytics':
Module 'Magento_Ui':
Module 'Magento_GroupedImportExport':
Module 'Magento_GroupedProduct':
Module 'Magento_DownloadableImportExport':
Module 'Magento_Vault':
Module 'Magento_Security':
Module 'Magento_LayeredNavigation':
Module 'Magento_Marketplace':
Module 'Magento_MediaStorage':
Module 'Magento_CatalogRule':
Module 'Magento_Multishipping':
Module 'Magento_ConfigurableProduct':
Module 'Magento_Newsletter':
Module 'Magento_OfflinePayments':
Module 'Magento_SalesRule':
Module 'Magento_PageCache':
Module 'Magento_Authorizenet':
Module 'Magento_Paypal':
Module 'Magento_Persistent':
Module 'Magento_ProductAlert':
Module 'Magento_ProductVideo':
Module 'Adyen_Payment':
Module 'Magento_Reports':
Module 'Magento_RequireJs':
Module 'Magento_Review':
Module 'Magento_Rss':
Module 'Magento_CatalogRuleConfigurable':
Module 'Magento_NewRelicReporting':
Module 'Magento_SalesInventory':
Module 'Magento_OfflineShipping':
Module 'Magento_Captcha':
Module 'Magento_SampleData':
Module 'Magento_CatalogSearch':
Module 'Magento_Integration':
Module 'Magento_SendFriend':
Module 'Magento_Shipping':
Module 'Magento_Sitemap':
Module 'J2t_Rewardpoints':
Module 'Magento_Swagger':
Module 'Magento_Swatches':
Module 'Magento_SwatchesLayeredNavigation':
Module 'Magento_Tax':
Module 'Magento_TaxImportExport':
Module 'Magento_GoogleAdwords':
Module 'Magento_Translation':
Module 'Magento_GoogleOptimizer':
Module 'Magento_Ups':
Module 'Magento_UrlRewrite':
Module 'Magento_EncryptionKey':
Module 'Magento_Usps':
Module 'Magento_Variable':
Module 'Magento_Braintree':
Module 'Magento_Version':
Module 'Magento_Webapi':
Module 'Magento_WebapiSecurity':
Module 'Magento_Weee':
Module 'Magento_CatalogWidget':
Module 'Magento_Wishlist':
Module 'Smile_ElasticsuiteCore':
Module 'Smile_ElasticsuiteCatalog':
Module 'Smile_ElasticsuiteCatalogRule':
Module 'Smile_ElasticsuiteCatalogOptimizer':
Module 'Smile_ElasticsuiteSwatches':
Module 'Smile_ElasticsuiteThesaurus':
Module 'Smile_ElasticsuiteTracker':
Module 'Smile_ElasticsuiteVirtualCategory':
Module 'WeltPixel_Backend':
Module 'WeltPixel_GoogleTagManager':
Schema post-updates:
Module 'Magento_Store':
Module 'CloudFlare_Plugin':
Module 'Magento_Directory':
Module 'Magento_Eav':
Module 'Magento_AdvancedPricingImportExport':
Module 'Magento_Backend':
Module 'Magento_Theme':
Module 'Magento_Authorization':
Module 'Magento_Backup':
Module 'Magento_Customer':
Module 'Magento_AdminNotification':
Module 'Magento_BundleImportExport':
Module 'Magento_CacheInvalidate':
Module 'Magento_Indexer':
Running schema recurring...
 [Magento\Framework\Exception\SessionException]                       
  Area code not set: Area code must be set before starting a session.  
                                                    
  [Magento\Framework\Exception\LocalizedException]  
  Area code is not set     
```
I'm running Magento ver. 2.1.4 on PHP 5.6 on CentOS7. I have done some research on this and it appears to have to do with the moment the area is set in the [class](http://magento.stackexchange.com/questions/96747/area-code-not-set-issue-in-custom-cli-commands-in-magento-2). 
Is there any way you could have a look at this?

**Comment by romainruaud** [comment]
Hello @juliangardnereurocios ,

we never faced such an issue when upgrading to new versions of the module.

Could you dig a bit more and check the full stack trace of the exception which is thrown ?

For me, this does not seem to be related to Elasticsuite but rather a Magento's bug as spoken in the thread you mention.

Regards

**Comment by juliangardnereurocios** [comment]
I can go through it more deeply, but isn't the solution suggested in the last comment?

> You shouldn't use setAreaCode in the __construct for CLI commands. When you run any command Magento collect and create instance for each script registered in your application. If there are more than one __construct with area code definition you will have the error.
> I suppose better to use the execute() method to set area code. Check the catalog module: vendor/magento/module-catalog/Console/Command/ImagesResizeCommand.php

**Comment by romainruaud** [comment]
I agree with your last answer. But the ElasticSuite module does not embed any CLI command. The setup:upgrade command you are launching is the Magento's native command.

We do not change the indexer command neither, the standard one is still used.

That's why I am thinking this is not an ElasticSuite issue but rather something due to Magento or to any other third party module installed on the projet.

Regards

**Comment by romainruaud** [comment]
I close this one due to be 2 weeks without answers and not clearly confirmed as being related to ElasticSuite.

Feel free to reopen it if needed.

Best regards,

**Comment by cykirsch** [comment]
Good morning, I ran into the same issue today and I just want to drop a note with the problem I found and the resolution.

The indexer loads up the StockItemRepository, which we had overridden and added to its constructor. And one of the dependencies injected into that new constructor eventually loaded up the session. The session, of course, requires the area code to be set. I used the Proxy solution found in the StackExchange question linked above. https://magento.stackexchange.com/a/162145/4297

Just had to add \Proxy to the Session class, so now the class doesn't load until it is needed. Since the install script never uses the session, it's never loaded and therefore doesn't throw an error for not having an area code:

before: `use Magento\Backend\Model\Auth\Session`
after: `use Magento\Backend\Model\Auth\Session\Proxy as Session;`
